### DISCUSSION TITLE
Combining dashboard and REST API?

### DISCUSSION THREAD

**Discussion by rkube** [GenieFramework/Genie.jl#686] [Q&A] [answered]
Hi,
I'm trying to develop a web app that combines a dashboard and also allows to query the data using a REST API.
I've followed the code from the [first dashboard tutorial](https://learn.genieframework.com/guides/first-dashboard)
and tried implementing the API like [here](https://genieframework.github.io/Genie.jl/dev/tutorials/7--Using_JSON_Payloads.html):

```
module App

using Main.genie_test
using GenieFramework
using PlotlyBase

# using Genie, Genie.Renderer.Json, Genie.Requests
# using HTTP

@genietools

@app begin
[...]
end

end

function ui()
[...]
end

@page("/", ui)
end


route("/test", method=POST) do 
    @show rawpayload()
    @show jsonpayload()
end

end
```

Then I start my dashboard with
```
julia> using GenieFramework
julia> Genie.loadapp()
julia> up()
```
and I see the dashboard in my browser.

When I ping the api with
```
$ curl http://localhost:8000/test
404 Not Found - /test. %       
```
I get a 404 error. How do I correctly combine the rest api into my dashboard?

**Comment by essenciary** [comment, marked as answer]
@rkube I think you've done it right. The issue seems to be that you defined "/test" over POST but you're making a GET request with curl.

**Comment by rkube** [comment]
Oh, right. Thanks!
I also had to include the route definition within `module App` . Then 
```
$ curl -X POST http://localhost:8000/test -H "Content-Type: application/json"  -d '{"id": 1234}'
```

shows the correct payloads.

**Comment by essenciary** [comment]
The routes are registered at app level, so they can be declared anywhere. But of course, you need to make sure that the file where you declare the routes is actually executed.
